Tests/SwiftSDKGeneratorTests/EndToEndTests.swift

```diff
@@ -49,66 +49,80 @@ extension FileManager {
-    let generatorOutput = try await Shell.readStdout(
-      "cd \(packageDirectory) && swift run --scratch-path \"\(scratchPath)\" swift-sdk-generator make-linux-sdk \(runArguments)"
-    )
-
-    let installCommand = try XCTUnwrap(generatorOutput.split(separator: "\n").first {
-      $0.contains("swift experimental-sdk install")
-    })
+// Building an SDK requires running the sdk-generator with `swift run swift-sdk-generator`.
+// This takes a lock on `.build`, but if the tests are being run by `swift test` the outer Swift Package Manager
+// instance will already hold this lock, causing the test to deadlock.   We can work around this by giving
+// the `swift run swift-sdk-generator` instance its own scratch directory.
+func buildSDK(_ logger: Logger, scratchPath: String, withArguments runArguments: String) async throws -> String {
+  var logger = logger
+  logger[metadataKey: "runArguments"] = "\"\(runArguments)\""
+  logger[metadataKey: "scratchPath"] = "\(scratchPath)"
+
+  logger.info("Building SDK")
+
+  var packageDirectory = FilePath(#filePath)
+  packageDirectory.removeLastComponent()
+  packageDirectory.removeLastComponent()
+
+  let generatorOutput = try await Shell.readStdout(
+    "cd \(packageDirectory) && swift run --scratch-path \"\(scratchPath)\" swift-sdk-generator make-linux-sdk \(runArguments)"
+  )
+  logger.info("Finished building SDK")
+
+  let installCommand = try XCTUnwrap(generatorOutput.split(separator: "\n").first {
+    $0.contains("swift experimental-sdk install")
+  })
+
+  let bundleName = try XCTUnwrap(
+    FilePath(String(XCTUnwrap(installCommand.split(separator: " ").last))).components.last
+  ).stem
+  logger[metadataKey: "bundleName"] = "\(bundleName)"
+
+  logger.info("Checking installed SDKs")
+  let installedSDKs = try await Shell.readStdout("swift experimental-sdk list").components(separatedBy: "\n")
+
+  // Make sure this bundle hasn't been installed already.
+  if installedSDKs.contains(bundleName) {
+    logger.info("Removing existing SDK")
+    try await Shell.run("swift experimental-sdk remove \(bundleName)")
+  }
 
-    let bundleName = try XCTUnwrap(
-      FilePath(String(XCTUnwrap(installCommand.split(separator: " ").last))).components.last
-    ).stem
+  logger.info("Installing new SDK")
+  let installOutput = try await Shell.readStdout(String(installCommand))
+  XCTAssertTrue(installOutput.contains("successfully installed"))
 
-    let installedSDKs = try await Shell.readStdout("swift experimental-sdk list").components(separatedBy: "\n")
+  return bundleName
+}
 
-    // Make sure this bundle hasn't been installed already.
-    if installedSDKs.contains(bundleName) {
-      try await Shell.run("swift experimental-sdk remove \(bundleName)")
+private let testcases = [
+  #"""
+    // Default program generated by swift package init
+    print("Hello, world!")
+  """#,
+  #"""
+    // Check that libc_nonshared.a is linked properly
+    import Foundation
```

Comment:
Not for this PR, but I think in the future we should test more of the core libraries here: FoundationEssentials, FoundationInternationalization, FoundationNetworking, FoundationXML, Dispatch, RegexBuilder, Synchronization, and Observation maybe? Since these are separate modules and are linked separately, we need to make sure that all necessary files are copied and discovered correctly.

Reply:
Created #157